### PR TITLE
x11: rework api to be multi-window capable

### PR DESCRIPTION
**Pull request by i509VCB**
This should allow the X11 backend to create multiple windows.

I have not implemented the handling of secondary windows in anvil yet but the single primary window is tested as working.

### CODE REVIEW

File: src/backend/x11/mod.rs [hunk 1]
```diff
@@ -441,7 +441,7 @@ impl X11Surface {
     /// This will fail if the backend has already been used to create a surface.
     pub fn new(
         backend: &mut X11Backend,
-        device: gbm::Device<DrmNode>,
+        device: Arc<Mutex<gbm::Device<DrmNode>>>,
```
Comment (Drakulix) [inline src/backend/x11/mod.rs:420]: I wonder, if this is something we should rather handle internally.
Although you can now share allocators between different backends with this PR, users should probably not be encouraged to do that and rather create multiple `gbm::Device`s for different backends.

File: src/backend/x11/mod.rs [hunk 2]
```diff
@@ -472,7 +472,7 @@ impl X11Surface {
     }
 
     /// Returns a handle to the GBM device used to allocate buffers.
-    pub fn device(&self) -> &gbm::Device<DrmNode> {
+    pub fn device(&self) -> &Arc<Mutex<gbm::Device<DrmNode>>> {
```
Comment (Drakulix) [inline src/backend/x11/mod.rs:469]: If we decide to hide the `Arc<Mutex<_>>` (see comment above), we could return a `MutexGuard<'_, gbm::Device<DrmNode>>` here instead.

**Review by Drakulix — COMMENTED** [review]
Looks good on first glance.

Do you plan to implement multi-window support for anvil in this PR or follow up with that later?

File: src/backend/x11/mod.rs (hunk 1 above)
Comment (i509VCB) [reply src/backend/x11/mod.rs:420]: Well the constructor for `X11Surface` needs to get the gbm device somehow. We don't make the X11Backend or X11Handle create the device at the moment.

To propagate the device to new X11Surfaces, we would need to either bootstrap off one surface or make an intermediate object for the device that internally holds the Mutex?

**Comment by i509VCB** [comment]
@Drakulix I will probably deal with multi-window in anvil at a later date because some of the input abstraction and windowed backend event handling is not yet configured for multiple outputs.

File: src/backend/x11/mod.rs (hunk 1 above)
Comment (Drakulix) [reply src/backend/x11/mod.rs:420]: I though about giving the `X11Backend` a `create_surface` function and making this constructor private.

Comment (i509VCB) [reply src/backend/x11/mod.rs:420]: Hmm that seems like a fine way around this issue

**Comment by Drakulix** [comment]
I think this PR is good as soon as CI agrees.

**Comment by i509VCB** [comment]
All the errors here should have been clippy update for 1.57, so rerun actions I guess?

**Comment by Drakulix** [comment]
> All the errors here should have been clippy update for 1.57, so rerun actions I guess? 

Or rebase onto master and push?

**Comment by i509VCB** [comment]
this appears to build now

**Pull request merged**